ledger-api/rs-grpc-akka/src/main/scala/com/digitalasset/grpc/adapter/server/akka/StreamingServiceLifecycleManagement.scala

```diff
@@ -58,4 +64,39 @@ trait StreamingServiceLifecycleManagement extends AutoCloseable {
       }
     }
   }
+
+  implicit class ScalaPbOptimizationsFlow[ScalaPbMessage <: GeneratedMessage with AnyRef, Mat](
+      val original: Source[ScalaPbMessage, Mat]
+  ) {
+
+    /** Optimization for gRPC stream throughput.
+      *
+      * gRPC internal logic marshalls the protobuf response payloads sequentially before
+      * sending them over the wire (see io.grpc.ServerCallImpl.sendMessageInternal), imposing as limit
+      * the maximum marshalling throughput of a payload type.
+      *
+      * We've observed empirically that ScalaPB-generated message classes have associated marshallers
+      * with significant latencies when encoding complex payloads (e.g. [[com.daml.ledger.api.v1.transaction_service.GetTransactionTreesResponse]]),
+      * with the gRPC marshalling bottleneck appearing in some performance tests.
+      *
+      * As an alleviation of the problem, we can leverage the fact that ScalaPB message classes have the serializedSize value memoized,
+      * (see [[scalapb.GeneratedMessage.writeTo]]), whose computation is roughly half of the entire marshalling step.
+      *
+      * This optimization method takes advantage of the memoized value and forces the message's serializedSize computation,
+      * roughly doubling the maximum theoretical ScalaPB stream throughput over the gRPC server layer.
+      *
+      * @return A new source with precomputed serializedSize for the [[scalapb.GeneratedMessage]]
+      */
+    def precomputeSerializedSize: Source[ScalaPbMessage, Mat] =
+      original.map { msg =>
+        // At ScalaPB 0.11.8, computation of serializedSize is thread-safe but the memoization mechanism
+        // which this optimization is relying on is not.
+        // Hence, synchronize on the message to ensure across-threads visibility of the memoized value,
```

Comment:
Out of interest and trying to understand it better:
1) Had Scala-PB made memoization thread safe (such that the memoized value is visible across threads), would we still need any optimizing being done on our side?
2) What threads are referred to in "... ensure across-threads ..." visibility; are these some threads that handle ScalaPB marshalling in parallel?

Reply:
each auto ScalaPB generated class has a `private[this] var __serializedSizeMemoized: _root_.scala.Int = 0` whose update is thread-safe (it can't be corrupted if it's update is triggered simultaneously by different threads). But it isn't `@volatile` . So, if we force its memoization on one thread, there is no guarantee that it will be visible to the thread calling marshalling, which could recompute again the value.

Reply:
Thanks!